server/resp/src/main/java/org/infinispan/server/resp/commands/generic/SORT.java

```diff
@@ -0,0 +1,242 @@
+
+import io.netty.channel.ChannelHandlerContext;
+import org.infinispan.AdvancedCache;
+import org.infinispan.commons.dataconversion.MediaType;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.multimap.impl.EmbeddedMultimapListCache;
+import org.infinispan.multimap.impl.ScoredValue;
+import org.infinispan.multimap.impl.SortableBucket;
+import org.infinispan.multimap.impl.internal.MultimapObjectWrapper;
+import org.infinispan.server.resp.Consumers;
+import org.infinispan.server.resp.Resp3Handler;
+import org.infinispan.server.resp.RespCommand;
+import org.infinispan.server.resp.RespErrorUtil;
+import org.infinispan.server.resp.RespRequestHandler;
+import org.infinispan.server.resp.commands.LimitArgument;
+import org.infinispan.server.resp.commands.Resp3Command;
+import org.infinispan.util.concurrent.CompletionStages;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Returns or stores the elements contained in the list, set or sorted set at key.
+ *
+ * There is also the {@link SORT_RO} read-only variant of this command.
+ *
+ * @since 15.0
+ * @see <a href="https://redis.io/commands/sort/">Redis Documentation</a>
+ */
+public class SORT extends RespCommand implements Resp3Command {
+   public static final char REPLACEMENT = '*';
+   public static final String VALUE_PATTERN = "#";
+   private boolean readonly = false;
+
+   public enum Arg {
+      ALPHA, ASC, DESC, LIMIT, STORE, BY, GET
+   }
+
+   public SORT() {
+      super(-2, 1, 1, 1);
+   }
+
+   public void disableStore() {
+      this.readonly = true;
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
+                                                      ChannelHandlerContext ctx,
+                                                      List<byte[]> arguments) {
+
+      int pos = 0;
+      byte[] key = arguments.get(pos++);
+      final SortableBucket.SortOptions sortOptions = new SortableBucket.SortOptions();
+      byte[] destination = null;
+      String pattern = null;
+      List<String> getObjectsPatterns = new ArrayList<>();
+      while (pos < arguments.size()) {
+         switch (SORT.Arg.valueOf(new String(arguments.get(pos++)).toUpperCase())) {
+            case ALPHA:
+               sortOptions.alpha = true;
+               break;
+            case ASC:
+               sortOptions.asc = true;
+               break;
+            case DESC:
+               sortOptions.asc = false;
+               break;
+            case STORE:
+               if (readonly) {
+                  RespErrorUtil.syntaxError(handler.allocator());
+                  return handler.myStage();
+               }
+               destination = arguments.get(pos++);
```

Comment:
Sorry, I meant that both the `STORE` and `BY` clauses do not verify if the index exists on the `arguments` list. The `GET` clause is doing the verification correctly. Otherwise, in case of syntax error:

```text
127.0.0.1:11222> RPUSH numbers 1 3 4 8 1 0 -1 19 -22 3
(integer) 10
127.0.0.1:11222> SORT numbers ALPHA DESC STORE
Error: Server closed the connection
not connected> 
```

We get:

```text
2023-10-03 16:35:19,400 WARN  [o.i.s.r.RespHandler] ISPN012003: Received an unexpected exception. java.lang.IndexOutOfBoundsException: Index 4 out of bounds for length 4
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:385)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at org.infinispan.server.resp.commands.generic.SORT.perform(SORT.java:80)
	at org.infinispan.server.resp.Resp3Handler.actualHandleRequest(Resp3Handler.java:75)
	at org.infinispan.server.resp.RespRequestHandler.handleRequest(RespRequestHandler.java:61)
	at org.infinispan.server.resp.RespHandler.handleCommandAndArguments(RespHandler.java:147)
	at org.infinispan.server.resp.RespHandler.channelRead(RespHandler.java:129)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at org.infinispan.server.core.transport.StatsChannelHandler.channelRead(StatsChannelHandler.java:28)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:499)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:397)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:1583)

```

Reply:
verification added